soc/xtensa/intel_adsp/ace/power.c

```diff
@@ -148,21 +182,78 @@ __asm__(".align 4\n\t"
 	"  add sp, sp, a2\n\t"
 	"  call0 power_gate_exit\n\t");
 
+__imr FUNC_NORETURN void pm_state_imr_restore(void)
+{
+	struct imr_layout *imr_layout = (struct imr_layout *)(IMR_LAYOUT_ADDRESS);
+	/* restore lpsram power and contents */
+	bmemcpy(z_soc_uncached_ptr(UINT_TO_POINTER(LP_SRAM_BASE)),
+		imr_layout->imr_state.header.imr_ram_storage,
+		LP_SRAM_SIZE);
+
+	/* restore HPSRAM contents, mapping and power states */
+	adsp_mm_restore_context(imr_layout->imr_state.header.imr_ram_storage+LP_SRAM_SIZE);
+
+	/* this function won't return, it will restore a saved state */
+	power_off_exit();
+}
+
 __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 {
 	ARG_UNUSED(substate_id);
 	uint32_t cpu = arch_proc_id();
 
 	if (state == PM_STATE_SOFT_OFF) {
+		/* save interrupt state and turn off all interrupts */
+		core_desc[cpu].intenable = XTENSA_RSR("INTENABLE");
+		z_xt_ints_off(0xffffffff);
+		core_desc[cpu].bctl = DFDSPBRCP.bootctl[cpu].bctl;
 		DFDSPBRCP.bootctl[cpu].wdtcs = DFDSPBRCP_WDT_RESTART_COMMAND;
 		DFDSPBRCP.bootctl[cpu].bctl &= ~DFDSPBRCP_BCTL_WAITIPCG;
 		soc_cpus_active[cpu] = false;
 		z_xtensa_cache_flush_inv_all();
 		if (cpu == 0) {
-			/* FIXME: this value should come from MM */
-			uint32_t hpsram_mask[1] = { 0x3FFFFF };
+			/* save storage and restore information to imr */
+			__ASSERT_NO_MSG(global_imr_ram_storage != NULL);
+			struct imr_layout *imr_layout = (struct imr_layout *)(IMR_LAYOUT_ADDRESS);
+
+			imr_layout->imr_state.header.adsp_imr_magic = ADSP_IMR_MAGIC_VALUE;
+			imr_layout->imr_state.header.imr_restore_vector =
+					(void *)rom_entry_d3_restore;
+			imr_layout->imr_state.header.imr_ram_storage = global_imr_ram_storage;
+			z_xtensa_cache_flush(imr_layout, sizeof(*imr_layout));
+
+			/* save CPU context here
+			 * when _restore_core_context() is called, it will return directly to
+			 * the caller of this procedure
+			 * any changes to CPU context after _save_core_context
+			 * will be lost when power_down is executed
+			 * Only data in the imr region survives
+			 */
+			xthal_window_spill();
+			_save_core_context(cpu);
+
+			/* save LPSRAM - a simple copy */
+			memcpy(global_imr_ram_storage, LP_SRAM_BASE, LP_SRAM_SIZE);
+
+			/* save HPSRAM - a multi step procedure, executed by a TLB driver
+			 * the TLB driver will change memory mapping
+			 * leaving the system not operational
+			 * it must be called directly here,
+			 * just before power_down
+			 */
+			const struct device *tlb_dev = DEVICE_DT_GET(DT_NODELABEL(tlb));
 
-			power_down(true, uncache_to_cache(&hpsram_mask[0]),
+			__ASSERT_NO_MSG(dev != NULL);
```

Comment:
Shouldn't it be `tlb_device` ?

Reply:
True! But that means - __ASSERT_NO_MSG macro is off and does not check anything

Reply:
@marcinszkudlinski Ack, work in progress to make asserts enabled in at least subset of our CI -> https://github.com/thesofproject/sof/pull/6530 -- first attempt failed, but will wbe looking at this again.